distributed/scheduler.py

```diff
@@ -3090,16 +3091,45 @@ def get_task_stream(self, comm=None, start=None, stop=None, count=None):
         return ts.collect(start=start, stop=stop, count=count)
 
     @gen.coroutine
-    def register_worker_callbacks(self, comm, setup=None):
-        """ Registers a setup function, and call it on every worker """
-        if setup is None:
-            raise gen.Return({})
+    def register_worker_callbacks(self, comm, callbacks, names):
```

Comment:
I preserved the Client API, but changed the scheduler API in a in compatible way:

This is to avoid
```
register_worker_callbacks(self, comm, setup, setup_name, teardown, teardown_name):
```

Reply:
I understand the problem, but I'm not entirely convinced by the proposed solution. I've not much to propose though...

Are you expecting that we register more than a setup and teardown callbacks at once in the future?

Could we use tuples instead of dict containing several callbacks? e.g stick to
```python 
register_worker_callbacks(self, comm, setup=None, teardown=None):
```

But where we expect `setup` or `teardown` to be `(name, function)`?

Reply:
I will rework the PR along this line.

Reply:
Sorry for not responding to this earlier.  I personally prefer asking for the `name=` keyword explicitly rather than passing it through as a tuple.  

```python
client.register_worker_callbacks(setup=setup_database, teardown=teardown_database, name='database')
```

I personally find passing tuples somewhat awkward.

@guillaumeeb what was your concern with the `name=` keyword solution?  That is it hard to provide ordered parameter compatibility going forward?

Reply:
I liked the tuple because:

- each callback needs a different name. 
- it was  decided `register_worker_callbacks` can register multiple callbacks in one call.
- there is no clean way to assign names to callbacks with a single name argument and multiple callbacks.

Reply:
My concern was not about the `name=` keyword solution, but about the `callbacks=dict` solution, which I find kind of obscure even if documented. The tuple may not be better though.

I think I prefer explicit `register_worker_callbacks(self, comm, setup, setup_name, teardown, teardown_name)` as the one with only a dict of callbacks.

Reply:
There wouldn't be a problem if the function was instead:

`register_worker_callback(self, comm, callback_type, function , name)`

I doubt there is need to make it atomic registering several callbacks.

Reply:
Here is another possibility. 

```
register_worker_callbacks(self, comm, setup, name=None)
```

If `setup` takes two arguments `setup(worker, event)`, then we will call it as 
`setup(worker, 'setup')`
`setup(worker, 'teardown')`.

If `setup` takes one or zero argument, we only call it during setup.

Reply:
Maybe when we add teardowns we decide that we need classes instead?

So today

```python
def register_worker_callbacks(self, comm, setup, name=None):
```

Where setup is a callable.

In the future

```python
def register_worker_callbacks(self, comm, cb, name=None):
```

Where cb is either a callable, in which case it is interpretted as a setup function, or it is a class with the following structure:

```python
class MyWorkerCallback:
    def __init__(self, worker):
        ...
    def setup(self):
        ...
    def tearDown(self):
        ...
    ...
```

In this way we continue to maintain the same structure today, provide a path towards supporting teardowns in the future, but don't have to rely on implicit tuple semantics.